### PR TITLE
Add helper to get errno and error message

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

Add platform aware helper to fetch errno message string.

### Motivation and Context

For usage in #20077

### CODE REVIEW

File: onnxruntime/core/platform/env.cc [hunk 1]
```diff
@@ -34,4 +34,29 @@ std::ostream& operator<<(std::ostream& os, gsl::span<const LogicalProcessors> af
...
+  if (err != 0) {
+    char buf[512];
+
+#if defined(_WIN32)
+    auto ret = strerror_s(buf, sizeof(buf), err);
+    msg = ret == 0 ? buf : "Failed to get error message";  // buf is guaranteed to be null terminated by strerror_s
+#else
+    // strerror_r return type differs by platform.
+    auto ret = strerror_r(err, buf, sizeof(buf));
+    if (std::is_same_v<decltype(ret), int>) {  // POSIX returns int
```
Comment (edgchen1) [inline onnxruntime/core/platform/env.cc:50]: nit: if constexpr
```suggestion
    if constexpr (std::is_same_v<decltype(ret), int>) {  // POSIX returns int
```

File: onnxruntime/test/platform/env_test.cc [hunk 2]
```diff
@@ -33,5 +33,13 @@ TEST(PlatformEnvTest, DirectoryCreationAndDeletion) {
   ASSERT_FALSE(env.FolderExists(root_dir));
 }
 
+TEST(PlatformEnvTest, GetErrnoInfo) {
+  // command that should generate an errno error
+  std::ifstream file("non_existent_file");
+  ASSERT_TRUE(file.fail());
+  auto [err, msg] = GetErrnoInfo();
+  ASSERT_NE(err, 0);
+  ASSERT_EQ(msg, "No such file or directory");
```
Comment (edgchen1) [inline onnxruntime/test/platform/env_test.cc:42]: the error message is locale specific. maybe check for err == ENOENT instead?

File: onnxruntime/core/platform/env.cc (hunk 1 above)
Comment (snnn) [reply onnxruntime/core/platform/env.cc:50]: On Linux it highly depends on whether the _GNU_SOURCE macro is defined or not. For example, whether we are using "--std=c++17" or "--std=gnu++17". It is really very tricky. Thanks for providing such a so great solution.

File: onnxruntime/test/platform/env_test.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/test/platform/env_test.cc:42]: I wanted to check that a string was set and was valid. Is there an `if` condition that could limit that to a specific locale?

File: onnxruntime/core/platform/env.cc (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/core/platform/env.cc:50]: This is the approach used in absl so I take no credit for it.

File: onnxruntime/test/platform/env_test.cc (hunk 2 above)
Comment (edgchen1) [reply onnxruntime/test/platform/env_test.cc:42]: hm, I think setlocale can query it. not sure what value we should test for though.
https://en.cppreference.com/w/cpp/locale/setlocale

maybe another approach is to explicitly set it to be a certain locale for this test, and then reset it to whatever it was before? e.g., the "C" locale.

Comment (edgchen1) [reply onnxruntime/test/platform/env_test.cc:42]: also, I think the error message is implementation-defined

**Review by edgchen1 — APPROVED** [review]

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**